source/plugin/event/causes.rst

```diff
@@ -69,55 +97,137 @@ object matching the provided type.
+Using the CauseStackManager
+===========================
+
+.. note::
+
+    The ``CauseStackManager`` only works on the main server thread. If you call it from a
+    different thread, an ``IllegalStateException`` will be thrown. Ensure you are on the main
+    server thread **before** calling methods on the ``CauseStackManager``.
+
+If you are creating your event on the main thread, then use the :javadoc:`CauseStackManager`, which can
+be found at :javadoc:`Sponge#getCauseStackManager()`. The ``CauseStackManager`` tracks the potential
+causes of events as the game runs, allowing for easy retrieval of the current ``Cause`` without effort.
+To see the current cause, call :javadoc:`CauseStackManager#getCurrentCause()`. You may notice that your
+plugin's :javadoc:`PluginContainer` is already in the returned ``Cause``, as plugins are one of the
+objects tracked by the manager. Using the ``CauseStackManager`` for creating causes removes the
+need for boilerplate-like code where you supply objects like your plugin container, you can concentrate
+on adding your own causes.
+
+Before adding your own causes, you should push a cause stack frame to the manager. Adding a frame acts
+as a saved state, when you are done with your causes, the removal of the frame returns the manager to
+its original state.
+
+For example, if you were to fire an event that was simulating another player in a sudo like command,
+you may want to add your acting player in the cause and the game profile of the player that you are
+simulating in the context (as the simulated player is not directly responsible for the event being fired.)
+
+**Creating a custom Cause with the CauseStackManager**
+
+.. code-block:: java
+
+    // In your code these would be populated
+    CommandSource sourceRunningSudo;
+    Player playerToSimulate;
+    try (CauseStackManager.StackFrame frame = Sponge.getCauseStackManager().pushCauseFrame()) {
+      // This would be the player that is causing the action (would have run the /sudo command)
+      frame.pushCause(sourceRunningSudo);
+
+      // You may want to push the simulated player so that it's at the root for compatibility
+      // with other plugins that listen for the direct cause
+      frame.pushCause(playerToSimulate);
+
+      // This would contain the GameProfile of the player to simulate, they are not a direct
+      // cause, but provide additional information to the event listeners so that they know
+      // the root player is a simulated one and may not have invoked the event.
+      frame.addContext(EventContextKeys.SIMULATED_PLAYER, playerToSimulate.getProfile());
+
+      // Add your event code here.
+      // You can get the current cause at any time using the getCurrentCause method
+      // (which is available on the fram as well as the CauseStackManager)
+      Cause cause = frame.getCurrentCause();
+    }
+
+Note that the last item you push to the cause stack will be the root of the ``Cause``.
```

Comment:
May want to link to the documentation of the Stack data structure to explain that it's a LIFO collection, as some developers may get confused about why ordering matters.